Get iers.org download to work

eteq opened the pull request:
This fixes #9499 - it adds a test that catches the original problem from #9499 and implements a combination of what @pllim and @StuartLittlefair suggested in #9499 of passing headers that apparently are required by the iers.org http server.

A subtle but important modification is that it allows `download_file` to pass in custom http headers.  I haven't directly tested that feature since this PR is focused on fixing #9499 and thereby tests it indirectly.  But I'm open to ideas about how one might test that (i.e. if there's some canonical web site that just returns in the result file the request headers or something like that...).

Note this should await merging until it can be rebased on #9182, because it quite plausible it will interact with it in possibly unexpected ways.

**EDIT -- NOTE TO REVIEWS: Check the numpy dev job that is allowed to fail because that same job deals with remote data tests.**

eteq commented:
There's also a question whether we think we need to backport this to 2.0.x or whatever.  I know we really don't want to, but given the current release just straight up doesn't work it might be necessary :cry:

bsipocz commented:
I'm 👎 on getting yet another release out. We need to get on with 4.0 instead.

eteq commented:
Rebased. It wasn't *that* bad in the end, @pllim.  But @aarchiba may want to give this PR a quick look to make sure I didn't miss anything important about how the new machinery was meant to be used.

Might be good to have @mhvk or @adrn have a look at the IERS changes, as the `SPECIAL_HTTP_HEADERS_NEEDED` approach is a bit hackier (but maybe not so much that's it's not ok...)

pllim commented on CHANGES.rst:
```diff
@@ -267,6 +267,9 @@ astropy.units
 astropy.utils
 ^^^^^^^^^^^^^
 
+- ``astropy.utils.data.download_file`` now provides an `http_headers` keyword to
```
This should be under "API change," not "new features," right?

Also, formatting typo:
```suggestion
- ``astropy.utils.data.download_file`` now provides an ``http_headers`` keyword to
```

pllim commented on astropy/utils/data.py:
```diff
@@ -1069,6 +1072,9 @@ def download_file(remote_url, cache=False, show_progress=True, timeout=None,
         ``~/.astropy/cache``.
 
+    http_headers : dict
+        Headers to pass into ``urlopen`` if needed.
```
```suggestion
        HTTP request headers to pass into ``urlopen`` if needed.
```

pllim commented on astropy/utils/iers/iers.py:
```diff
@@ -43,10 +43,16 @@
 
+# maps specific download urls to the additional http headers needed for them -
+# populated where specific urls are given below.
+SPECIAL_HTTP_HEADERS_NEEDED = {}
```
Is this even necessary? Can't you pass that in blindly regardless of URL? What if user change the value of `IERS_A_URL_MIRROR` at run time?

eteq replied:
My hypothesis was that other custom headers might be called for, but as you an @aarchiba pointed out maybe it's best to always do it...

aarchiba commented:
So this is to fix the issue that the IERS server requires a User-Agent and an Accepts header? It seems like ``requests`` always passes these headers and so just works - is there any reason we shouldn't just always pass these headers?

eteq commented:
I think I've addressed the comments, but also went with @pllim and @aarchiba's suggestions of *always* sending ``Accept`` and ``User-Agent`` headers. I'm a bit worried that this will have unintended consequences so tried to make all that by itself in a couple of easily-revertable commits, but I'm OK with living dangerous :TM: since as @aarchiba said it seems like it really *should* work (and is being a better web citizen anyway...)

pllim commented:
@eteq , what about `get_readable_fileobj`, `get_pkg_data_*`, and `get_file_contents` in `data.py`?

Maybe @aarchiba can clarify, but after #9182, `download_file` now use `sources` keyword to handle mirrors. I am afraid the way it is done now is not working as it is supposed to be anymore; i.e., if main IERS URL is not cached but its mirror is, `iers.py` will attempt to download from the main URL anyway and ignore the mirror's cache.

eteq commented:
@pllim - I see your point but I'm concerned about making a big change on that at this late of a stage, as I'm not confident that the time-out machinery in IERS would play as well with the sources approach. (Possibly due to ignorance, but still). So perhaps we should merge this so that things are back to working again, and then have a follow-on to properly uses ``sources`` in IERS?

eteq commented:
(Put another way, my thinking here is that this PR is focused on the minimum change that will get the IERS downloads to work again, since that's something of a 🔥for branching.)

aarchiba commented on astropy/utils/data.py:
```diff
@@ -1015,7 +1019,7 @@ def _download_file_from_source(source_url, show_progress=True, timeout=None,
 
 def download_file(remote_url, cache=False, show_progress=True, timeout=None,
-                  sources=None, pkgname='astropy'):
+                  sources=None, pkgname='astropy', http_headers={}):
```
Preferable to use None here rather than a mutable default argument; it's a customary way to avoid that pitfall.

aarchiba commented on astropy/utils/data.py:
```diff
@@ -1069,6 +1073,10 @@ def download_file(remote_url, cache=False, show_progress=True, timeout=None,
 
+    http_headers : dict
+        HTTP request headers to pass into ``urlopen`` if needed. (These headers
+        are ignored if the ``remote_url`` is not http.)
```
Actually it's `sources` that matters, not `remote_url`.

eteq replied:
Point taken, although it's confusing to work because `remote_url` does matter *if* `sources` is None.  Will try to wordsmith a bit.

aarchiba commented on astropy/utils/iers/iers.py:
```diff
@@ -90,6 +90,10 @@ def download_file(*args, **kwargs):
     ``**kwargs`` after temporarily setting the download_file remote timeout to
     the local ``iers.conf.remote_timeout`` value.
     """
+    url = kwargs.get('remote_url', args[0] if len(args) > 0 else None)
```
Where is this used?

eteq replied:
Oops!  Good catch, thanks @aarchiba  - that was left over from before the rebase. Will remove.

pllim commented on astropy/utils/data.py:
```diff
@@ -952,12 +952,16 @@ def check_free_space_in_dir(path, size):
 def _download_file_from_source(source_url, show_progress=True, timeout=None,
-                               remote_url=None, cache=False, pkgname='astropy'):
+                               remote_url=None, cache=False, pkgname='astropy',
+                               http_headers={'User-Agent': 'astropy',
```
I wasn't expecting this custom header to be the default all the way down here.

aarchiba replied:
For this internal function called only from one place there isn't actually a need for defaults at all, in fact it probably shouldn't have them (so as to catch calls that omit an argument); they could also be keyword-only. That's my fault. But passing a mutable default argument is bad form.

eteq replied:
Good point. I moved this construct to `download_files`. Still the relevant question of whether it's *desirable*, though.

pllim replied:
I propose set `http_headers` to `None` in `download_file`. It should only be set in `iers.py` because it is necessary there. The rest should be done in a case-by-case basis from higher level calls.

eteq commented:
I think I've addressed @pllim and @aarchiba's comments now, including adding `get_readable_fileobj` (but *not* package data because it seemed out-of-scope) - look good to you?

The one dangling thread still is whether we wanted the user-agent and accept headers for *everything* vs just IERS.  I tend toward the former because 1. we can merge it now and 2. that's actually what one is *supposed* to do, as demonstrated by requests.

pllim commented on astropy/utils/data.py:
```diff
@@ -1089,6 +1105,8 @@ def download_file(remote_url, cache=False, show_progress=True, timeout=None,
     if sources is None:
         sources = [remote_url]
+    if http_headers is None:
+        http_headers = {'User-Agent': 'astropy','Accept': '*/*'}
```
I still don't like this but I'll defer to @aarchiba for a second opinion.

eteq replied:
So just to be clear, @pllim, you're proposing instead having this in the iers.py `download_file` instead of this one?

I'm open to that but I think this is actually better because it will make it possible to keep better track of Astropy on remote services for the long-term (and my understanding is that you're "supposed" to do this according to some HTTP recommendations, but could be wrong about that).  So I agree @aarchiba is the deciding vote here.

aarchiba replied:
Sorry to be slow getting back to you on this, I wanted to read more about what good citizenship means in terms of HTTP headers. What is the correct User-Agent for a (semi-)unattended script? Is the Accepts claiming something peculiar? But the fact that `requests` apparently sends these headers by default on every request leads me to think that that is probably a good idea.

pllim replied:
So, a third-party package that is using `astropy.utils.data` to download files would be telling the service provider that the request come from `astropy`. Certainly, they can pass in their own header but I have a feeling that 95% of the people would just use the default. Is this side effect desirable?

adrn replied:
I think I'm fine with the current behavior as long as it is made very clear in the docstrings (right now they do not clearly expose what the default behavior is). 

As another idea, what if we made a config item for "package_user_agent" or something, and set it to "astropy" for astropy core? Then the default can be "None" or no user-agent name, but we give other people that use this functionality within other packages the ability to easily control their user-agent name?

adrn replied:
I'm coming around to the config item idea being the better solution, however, in the interest of time and given that things around this are broken in master, I would be fine with making an issue for that and merging this as is (after adding the default behavior to the docstring).

eteq commented:
@pllim or @bsipocz - one of you probably should approve this before we can merge it (since I can't as the author)

bsipocz commented:
> @pllim or @bsipocz - one of you probably should approve this before we can merge it (since I can't as the author)

I leave that to @pllim as I didn't at all followed this PR.

pllim commented:
I am hesitant to approve until we can resolve the headers conversation.

pllim commented on astropy/utils/data.py:
```diff
@@ -190,6 +190,11 @@ def get_readable_fileobj(name_or_obj, encoding=None, cache=False,
+    http_headers : dict
+        HTTP request headers to pass into ``urlopen`` if needed. (These headers
+        are ignored if the protocol for the ``name_or_obj``/``sources`` entry
+        is not a remote http URL.)
```
```suggestion
        is not a remote HTTP URL.)
```

pllim commented on astropy/utils/data.py:
```diff
@@ -1069,6 +1080,11 @@ def download_file(remote_url, cache=False, show_progress=True, timeout=None,
+    http_headers : dict
+        HTTP request headers to pass into ``urlopen`` if needed. (These headers
+        are ignored if the protocol for the ``remote_url``/``sources`` entry
+        is not http.)
```
```suggestion
        is not HTTP.)
```

pllim commented:
The remote data test failed with FTP timeout... related? https://travis-ci.org/astropy/astropy/jobs/608290503

eteq commented:
@pllim - hmm, lets see if the build from the minor docstring fix (which @adrn suggested to me verbally) also shows it.  I'm speculating it was just the standard temporary server issue but we will find out...  (Hopefully the sources trick will help too)

pllim commented on astropy/utils/data.py:
```diff
@@ -190,6 +190,12 @@ def get_readable_fileobj(name_or_obj, encoding=None, cache=False,
+        HTTP request headers to pass into ``urlopen`` if needed. (These headers
+        are ignored if the protocol for the ``name_or_obj``/``sources`` entry
+        is not a remote HTTP URL.) In the default case (None), the headers are
+        "User-Agent: astropy" and "Accept: */*".
```
I am mixing up RST with MarkDown... Does `*` mean anything special in rendering? Does this docstring need to be a raw string? Please check `giles`.

eteq replied:
Good point.  Switched to just using ``` `` ```literals

pllim commented:
Same timeout error as before. Does it pass when you run with `--remote-data` locally?

eteq commented:
Yes, the remote data tests are all working for me locally @pllim... very mysterious

pllim commented:
It could also be that the FTP service has blacklisted Travis CI as a spammer. 🤷‍♀  If it is persistent, we might need to `xfail` or `skip` it in the CI environment.

pllim commented:
`urllib.error.URLError: <urlopen error ftp error: timeout('timed out')>` is still there. Hmmm...

eteq commented:
Re: @pllim's comment (which was referencing the numpy-dev Travis build) - I cannot reproduce that. I tried the ``--remote-data`` option both locally and on CircleCI and everything works flawlessly.  @pllim out-of-band suggested it might be that the server is blocking Travis because of too many requests. So under that hypothesis I'm doing a skip-if-on-travis.  Assuming that makes things happy I'll make an issue to figure out why that's happening on travis and then we can merge this, since we're then still fixing master for users.

pllim commented:
For the sake of progress, if numpy-dev job passes this time, go ahead and merge it. I already approved.

pllim commented:
Doh... 🤦‍♀ 

```
______ test_iers_a_dl[https://datacenter.iers.org/data/9/finals2000A.all] ______
[XPASS(strict)] condition: TRAVIS
________________________________ test_iers_b_dl ________________________________
[XPASS(strict)] condition: TRAVIS
===== 2 failed, 14273 passed, 116 skipped, 92 xfailed in 445.61s (0:07:25) =====
```

pllim commented on astropy/utils/iers/tests/test_iers.py:
```diff
@@ -325,3 +328,28 @@ def test_IERS_B_parameters_loading_into_IERS_Auto():
+        iers.IERS_A.close()
+
+
+@pytest.mark.xfail('TRAVIS')
```
This wasn't failing in Travis, right? I only remember one failed and that is the FTP service.
```suggestion
```

pllim commented on astropy/utils/iers/tests/test_iers.py:
```diff
@@ -325,3 +328,28 @@ def test_IERS_B_parameters_loading_into_IERS_Auto():
+
+
+@pytest.mark.parametrize('iersa_url', [iers.IERS_A_URL, iers.IERS_A_URL_MIRROR])
+@pytest.mark.xfail('TRAVIS')
```
This needs to be set up such that it `xfail` only for `iers.IERS_A_URL` (the FTP service) and not the HTTP mirror. I can think of two ways:

1. Move the `xfail` to a `if` block inside the test function. But I don't think `xfail_strict` would catch it if it starts to `xpass` again some time in the future.
2. Turn this into a test class. Make the common code some hidden method that `pytest` won't collect. Then write two test methods wrappers around it, one with `xfail` marker and one without.

bsipocz commented:
Addressed last review and rebased to pick  up the mpldev fixes from master. We should merge if CI is green (after checking that the remote tests are passing, too!!)

bsipocz commented:
Thanks @eteq!
